### ISSUE TITLE
About "synth.audio-channels" usefulness  ?

### ISSUE THREAD

**Issue by jjceresa** [FluidSynth/fluidsynth#660]
This is not issue, but a more technical discussion about the real usefulness of "synth.audio-channels".

1) Actually, both `"synth.audio-groups"` and `"synth.audio-channels"` contribute to define the maximum number of mixer output buffer:
https://github.com/FluidSynth/fluidsynth/blob/31c4e12ac9f2f4eadfb6f034845e51bff7faf223/src/synth/fluid_synth.c#L755
```
nbuf = synth->audio_channels;
if(synth->audio_groups > nbuf)
{
        nbuf = synth->audio_groups;
}
```

2) The voice mapping (voice audio dry to buffer) is making use of `synth->audio_groups`
https://github.com/FluidSynth/fluidsynth/blob/31c4e12ac9f2f4eadfb6f034845e51bff7faf223/src/synth/fluid_voice.c#L360

3.1)It seems that `synth.audio-groups` always override `synth.audio-channels`, so what could be the real real usefulness of "synth.audio-channels" ? 
In other words, is there a case where `synth.audio-channels` is used independently of `synth.audio-groups` (apart that it is used by the obsolete fluid_synth_nwrite_float()) ?

**Comment by mawe42** [comment]
> In other words, is there a case where synth.audio-channels is used independently of synth.audio-groups (apart that it is used by the obsolete fluid_synth_nwrite_float()) ?

Yes, the jack driver makes direct use of the `synth.audio-channels` setting to determine the number of output channels.

**Comment by jjceresa** [comment]
>Yes, the jack driver makes direct use of the synth.audio-channels setting to determine the number of output channels.

Thanks Marcus, so I had missed this case in my search command !.

**However, it seems there is something wrong**. Let this case:  `"synth.audio-channels"`  > "synth.audio-groups"
- Jack driver expect that `synth->audio_channels` will be played to the audio card, but
- only `synth->audio_groups` will be mapped (see point 2 in previous comment) . This leads to some number of output on the audio card (synth->audio_channels - synth->audio_groups) will be silent, which is an issue for the user.

**Comment by jjceresa** [comment]
>Yes, the jack driver makes direct use of the synth.audio-channels setting to determine the number of output channels.

Thanks Marcus, so I had missed this case in my search command !.

**However, it seems there is something wrong**. Let this case:  `"synth.audio-channels"`  > "synth.audio-groups"
- Jack driver expect that `synth->audio_channels` will be played to the audio card, but
- only `synth->audio_groups` will be mapped (see point 2 in previous comment) . This leads to some number of output on the audio card (synth->audio_channels - synth->audio_groups) will be silent, which is an issue for the user.

**Comment by jjceresa** [comment]
>Yes, the jack driver makes direct use of the synth.audio-channels setting to determine the number of output channels.

Thanks Marcus, so I had missed this case in my search command !.

**However, it seems there is something wrong**. Let this case:  `"synth.audio-channels"`  > "synth.audio-groups"
- Jack driver expect that `synth->audio_channels` will be played to the audio card, but
- only `synth->audio_groups` will be mapped (see point 2 in previous comment) . This leads to some number of output on the audio card (synth->audio_channels - synth->audio_groups) will be silent, which is an issue for the user.

**Comment by jjceresa** [comment]
>Yes, the jack driver makes direct use of the synth.audio-channels setting to determine the number of output channels.

Thanks Marcus, so I had missed this case in my search command !.

**However, it seems there is something wrong**. Let this case:  `"synth.audio-channels"`  > "synth.audio-groups"
- Jack driver expect that `synth->audio_channels` will be played to the audio card, but
- only `synth->audio_groups` will be mapped (see point 2 in previous comment) . This leads to some number of output on the audio card (synth->audio_channels - synth->audio_groups) beeing silent, which is an issue for the user.

**Comment by jjceresa** [comment]
>Yes, the jack driver makes direct use of the synth.audio-channels setting to determine the number of output channels.

Thanks Marcus, so I had missed this case in my search command !.

**However, it seems there is something wrong**. Let this case:  `"synth.audio-channels"`  > "synth.audio-groups"
- Jack driver expect that `synth->audio_channels` will be played to the audio card, but
- only `synth->audio_groups` will be mapped (see point 2 in previous comment) . This leads to some number of output on the audio card (synth->audio_channels - synth->audio_groups) beeing silent, which is an issue for the user.

**Comment by jjceresa** [comment]
>Yes, the jack driver makes direct use of the synth.audio-channels setting to determine the number of output channels.

Thanks Marcus, so I had missed this case in my search command !.

**However, it seems there is something wrong**. Let this case:  `"synth.audio-channels"`  > "synth.audio-groups"
- Jack driver expect that `synth->audio_channels` will be played to the audio card, but
- only `synth->audio_groups` will be mapped (see point 2 in previous comment) . This leads to some number of output on the audio card (synth->audio_channels - synth->audio_groups)  silent, which is an issue for the user.

**Comment by jjceresa** [comment]
>Yes, the jack driver makes direct use of the synth.audio-channels setting to determine the number of output channels.

Thanks Marcus, so I had missed this case in my search command !.

**However, it seems there is something wrong**. Let this case:  `"synth.audio-channels"`  > "synth.audio-groups"
- Jack driver expect that `synth->audio_channels` will be played to the audio card, but
- only `synth->audio_groups` will be mapped (see point 2 in previous comment) . This leads to some number of output on the audio card (synth->audio_channels - synth->audio_groups)  silent, which is an issue for the user.

**Comment by jjceresa** [comment]
>Yes, the jack driver makes direct use of the synth.audio-channels setting to determine the number of output channels.

Thanks Marcus, so I had missed this case in my search command !.

**However, it seems there is something wrong**. Let this case:  `"synth.audio-channels"`  > "synth.audio-groups"
- Jack driver expect that `synth->audio_channels` will be played to the audio card, but
- only `synth->audio_groups` will be mapped (see point 2 in previous comment) . This leads to some number of output on the audio card (synth->audio_channels - synth->audio_groups)  silent, which is an issue for the user.

**Comment by jjceresa** [comment]
Sorry, for the duplicate. Sometimes GitHub behaves weird.

**Comment by derselbst** [comment]
>  `"synth.audio-channels"` > "synth.audio-groups"
> leads to some number of output on the audio card will be silent, which is an issue for the user.

That's the intended behaviour, I'd say.

I'll keep thinking about in which case synth.audio-channels makes sense.

**Comment by jjceresa** [comment]
>I'll keep thinking about in which case synth.audio-channels makes sense.

Just some notes about how I think this from the user point of view :
1)One important thing is the setting on which is based the MIDI channel `voice-mapping-to-audio-dry-output`.
   actually this setting is "synth.audio-groups". For the user that means he can expect to distribute all MIDI channels to
   the number of mixer's internal output indicated by this settings (i.e synth.audio-groups = 2 . The user see out0, out1).
   Actually, the mapping is hard coded, but later we could offer a flexible simple mapping support which allows the
   user to map  any MIDI channel to any of these internal output (out0,out1) . So the user should be aware of this
   internal `mixer mapping  capabilities`  regardless of the `audio driver capabilities`.
   
   At this stage it is obvious that the number of allocated internal dry buffers should be only dependent of  `synth.audio-groups`.
   (Note: Actually, the number allocated dry buffers is  `max (synth.audio-groups, synth.audio-channels)`. see point 1
    in first comment. In the case of  `synth.audio-channels > synth.audio-groups`, there were some buffers never rendered
   (synth.audio-channels -  synth.audio-groups) but still allocated which is a memory waste).

2) Multiple output audio driver (fluid-audio-driver-x + host-device-driver) capabilities depend both of `fluidsynth mixer` and `audio device hardware` respective capabilities. For examples: 
2.1) if this `fluid-audio-driver-x` look for  the audio device's channels limits (i.e 8 channels max).
   this `fluid-audio-driver-x` is  able to play the 2 outputs provided by the `fluidsynth internal mixer`.
2.2) if the user set `synth.audio-groups` to 3 or 4,  `fluid-audio-driver-x` is still not limited.
2.3) if the user set `synth.audio-groups` to 5,  `fluid-audio-driver-x` is able to see that it is limited by the audio device's `channels limit` (8 channels). The driver warn the user and fails.
In all case  `fluid-audio-driver-x look` for 'synth.audio-groups' settings and device's `channels limit`.

If `fluid-audio-driver-x` is jack it should behave similarly.

**Comment by jjceresa** [comment]
>That's the intended behaviour, I'd say.

Really!. So for example, the goal would be having the user able to "mute" some output ?.

**Comment by jjceresa** [comment]
>jjc said: is there a case where synth.audio-channels is used independently of synth.audio-groups

>Marcus said: Yes, the jack driver makes direct use of the synth.audio-channels setting to determine the number of output channels.

An other case, `fluid_synth_process_LOCAL()` is indirectly making used of  `synth.audio-channels` here:

https://github.com/FluidSynth/fluidsynth/blob/4b83ee55c0002ec8512f56506e237b8556009985/src/synth/fluid_synth.c#L3738

But I think `synth->audio-channels` should be replaced by `synth->audio-groups`.

**Comment by derselbst** [comment]
>  In the case of synth.audio-channels > synth.audio-groups, there were some buffers never rendered (synth.audio-channels - synth.audio-groups) but still allocated which is a memory waste).

If you use LADSPA, it might be intended to have some "silent" audio channels, because they would allow you to apply custom effects on these channels.

If not using LADSPA,  it would be a waste of memory, you are correct.

> But I think synth->audio-channels should be replaced by synth->audio-groups.

If the user uses `fluid_synth_process()` and additionally uses LADSPA in the way I explained above, these LADSPA effect channels must be rendered to the user provided buffers as well. Changing it to `synth->audio-groups` would not work in this case.

**Comment by jjceresa** [comment]
>If you use LADSPA, it might be intended to have some "silent" audio channels, because they would allow you to apply custom effects on these channels.

In the case of synth.audio_channels > synth.audio_groups, (i.e synth.audio_channels =4, synth->audio_groups = 2)
MIDI input will be rendered to buffers `out0,out1`, buffers `out2,out3` will be silent.

I have understood that LADSPA  read only `synth->audio_groups`so it have knowledge only of buffers `out0,out1` which are potential input/output for LADSPA.   @mawe42  How LADSPA could output effects on `out2,out3` ?

**Comment by derselbst** [comment]
I thought about buffer routing as explained by Marcus' reverb example here: https://lists.nongnu.org/archive/html/fluid-dev/2020-08/msg00001.html

However, to my surprise, you are correct, JJC. LADSPA only has `audio_groups` buffers. So the use-case I explained is impossible to achieve.

https://github.com/FluidSynth/fluidsynth/blob/4b83ee55c0002ec8512f56506e237b8556009985/src/synth/fluid_synth.c#L830-L831

In this case, `synth.audio-channels` seems pretty useless to me, as it de facto must always be equal to audio-groups.

**Comment by jjceresa** [comment]
>However, to my surprise, you are correct, JJC. LADSPA only has audio_groups buffers. So the use-case I explained is impossible to achieve.

Thanks Tom for confirming this. As I am not yet a LADSPA user, I was not sure about my understanding.

However,  @mawe42 opinion is welcome, because Marcus wrote this in LADSPA documentation https://github.com/FluidSynth/fluidsynth/blob/master/doc/ladspa.md  Section **Multi-Channel Output** :

"If you want all main ports to act as outputs as well as inputs to the effects, then you also need to increase the `synth.audio-channels setting`".

Starting from default settings values, please why LADSPA should need `synth.audio-channels` to be increased if `synth.audio-group` is set to 2 ?.

**Comment by mawe42** [comment]
My understand is that `audio-groups` determines how many internal buffers we use for rendering MIDI to audio. And that `audio-channels` is used by the audio drivers to decide which internal buffers should be available as output channels.

If you have `audio-groups` > `audio-channels`, you need to mix the extra groups into lower channels via LADSPA effects, otherwise the extra groups will not be heard.

If `audio-groups` = `audio-channels`, then all groups are available as output channels without manual mixing.

And `audio-groups` <  `audio-channels` does not make any sense, because you can't use the extra output channels.

All assuming you are using the jack driver, which is the only one supporting multi-channel output. For all other drivers, the audio-channels setting has not effect.

Have I misunderstood how audio-groups and audio-channels work?

**Comment by derselbst** [comment]
> Have I misunderstood how audio-groups and audio-channels work?

No, I think your explanations are correct.

 I just don't know whether `audio-groups` > `audio-channels` is actually a valid use-case though. IMO, this would only make sense if one sets `audio-groups==16`, so that each MIDI channel gets rendered to its dedicated internal buffer, while setting `audio-channels` to something else, e.g. `==2`. That way one would be able to cherry-pick each and every MIDI channel to either out channel 1 or out channel 2. But... why would you want to do something like that?

**Comment by jjceresa** [comment]
>Have I misunderstood how audio-groups and audio-channels work?

Your understanding is coherent with actual behaviour. 

So actually `audio-channels` is rather a jack specific setting use to restrict which internal buffers (in the range `[1..audio-groups] `) should be available as output channels.

Having `audio-groups > audio-channels` make the buffers in the range `]audio-channels..audio-groups]` not audible for jack only.
But these buffers could be audible using another driver (which doesn't yet exist).

Do you think that having buffers in the range  `]audio-channels..audio-groups]` not audible  should be a useful future for LADSPA users ?

And `audio-groups < audio-channels` does not make any sense, because buffers in the range `]audio-groups..audio-channels]` are not mapped by the mixer. However these buffers are actually allocated which is an error (see point 1 in first comment) .

**Comment by jjceresa** [comment]
>Marcus said. My understanding is that **audio-groups** determines how many internal buffers we use for rendering MIDI to audio. And that **audio-channels** is used by the audio drivers to decide which internal buffers should be available as output channels. If you have audio-groups > audio-channels, you need to mix the extra groups into lower channels via LADSPA effects, otherwise the extra groups will not be heard.

>Tom said: I just don't know whether audio-groups > audio-channels is actually a valid use-case though

> JJC said: Having audio-groups > audio-channels, do you think that having buffers in the range ]audio-channels..audio-groups] not audible should be a useful future for LADSPA users ?

I just realized a real use case for `audio-channels < audio-groups`. The user should think like that:
1)First (as Marcus said) the user should set  `audio-groups` in the intend to render MIDI channel in distinct buffers.
   This setting makes buffers in the range `[1..audio-groups]` potentially audible.
2)Typically `audio driver` should used only `audio-groups` and request the `audio device` to play the range `[1..audio-groups]` to
   distinct `device audio channel`. This will work only if the `audio device` has enough audio channels on board.
   In the case of  this `audio device` has number of device channel less then `audio-groups` , the user should diminish
   `audio-groups` setting.
3)But  this audio driver could also use only 'audio-channels' setting (instead of 'audio-groups') to select buffer .
3.1)Typically this `audio-channels` setting should be set to the same value than 'audio-groups' setting.
3.2)When using LADSPA, the case `audio-channels` < 'audio-groups' makes the range of potential audible buffers 
       [1..audio-groups] split in two sub ranges:  `lower-range`=[1..audio-channels]  `upper-range`=]audio-channels..audio-groups].
       Having MIDI channels audio separated in `upper-range` buffers makes possible to apply LADSPA effect only to these MIDI
       channels by routing `LADSPA input` to `upper-range` and  `LADSPA output` to `lower-range`.
       This way allows the user to get a better control on which MIDI channel a LADSPA effect should be applied.
       This is particularly useful when the audio device has a low number of device channels.

**Comment by jjceresa** [comment]
@mawe42, @derselbst. Thanks you for yours comments.

**Comment by jjceresa** [comment]
Will close this.

**Comment by jjceresa** [comment]
Cannot close this "issue". The "close and comment" button does not work !.

**Comment by mawe42** [comment]
Strange! I could close it...

**Comment by mawe42** [comment]
Maybe you need a more modern browser... ;-)

**Comment by jjceresa** [comment]
>Maybe you need a more modern browser... ;-)

Unfortunately modern browser doesn't install on Windows XP !. 
This worked well since the beginning of the year. Now since one week  I cannot do some actions that worked well before.
They have changed somethings. This isn't important because I am not fan of the GUI interface proposed by GitHub.
